src/app/AttributePathExpandIterator.h

```diff
@@ -20,81 +20,104 @@
+///      }
+///
+/// Usage requirements and assumptions:
+///
+///    - An ` AttributePathExpandIterator::Position` can only be used by a single AttributePathExpandIterator at a time.
+///
+///    - `position` is automatically updated by the AttributePathExpandIterator, so
+///      calling `Next` on the iterator will update the position cursor variable.
+///
 class AttributePathExpandIterator
 {
 public:
-    AttributePathExpandIterator(DataModel::Provider * provider, SingleLinkedListNode<AttributePathParams> * attributePath);
-
-    /**
-     * Proceed the iterator to the next attribute path in the given cluster info.
-     *
-     * Returns false if AttributePathExpandIteratorDataModeDataModel has exhausted all paths in the given AttributePathParams list.
-     */
-    bool Next();
-
-    /**
-     * Fills the aPath with the path the iterator currently points to.
-     * Returns false if the iterator is not pointing to a valid path (i.e. it has exhausted the cluster info).
-     */
-    bool Get(ConcreteAttributePath & aPath)
+    class Position
     {
-        aPath = mOutputPath;
-        return (mpAttributePath != nullptr);
-    }
+    public:
+        // Position is treated as a direct member access by the AttributePathExpandIterator, however it is opaque (except copying)
+        // for external code. We allow friendship here to not have specific get/set for methods (clearer interface and less
+        // likelihood of extra code usage).
+        friend class AttributePathExpandIterator;
 
-    /**
-     * Reset the iterator to the beginning of current cluster if we are in the middle of expanding a wildcard attribute id for some
-     * cluster.
-     *
-     * When attributes are changed in the middle of expanding a wildcard attribute, we need to reset the iterator, to provide the
-     * client with a consistent state of the cluster.
-     */
-    void ResetCurrentCluster();
-
-    /** Start iterating over the given `paths` */
-    inline void ResetTo(SingleLinkedListNode<AttributePathParams> * paths)
-    {
-        *this = AttributePathExpandIterator(mDataModelProvider, paths);
-    }
+        /// External callers can only ever start iterating on a new path from the beginning
+        static Position StartIterating(SingleLinkedListNode<AttributePathParams> * path) { return Position(path); }
+
+        /// Copies are allowed
+        Position(const Position &)             = default;
+        Position & operator=(const Position &) = default;
+
+        Position() : mAttributePath(nullptr) {}
+
+        /// Reset the iterator to the beginning of current cluster if we are in the middle of expanding a wildcard attribute id for
+        /// some cluster.
+        ///
+        /// When attributes are changed in the middle of expanding a wildcard attribute, we need to reset the iterator, to provide
+        /// the client with a consistent state of the cluster.
+        void IterateFromTheStartOfTheCurrentCluster()
```

Comment:
```suggestion
        void RestartClusterIterationIfAttributeWildcard()
```
or something?  The conditionality of the behavior is not reflected in the name....

Reply:
I am torn here - the optionality here is mainly because there is nothing to restart if no wildcard. I generally dislike methods with `And` or `If` in the name as they are hard to follow. 

How strongly do you feel that this is a problem?

Reply:
Renamed to get it through the review. However the length of this method name seems to indicate that we still have a design problem. I should not need to write long method names with if/then (the caller should have an if in case it is really needed...)